### PR TITLE
Timestamp messages for client in chat view

### PR DESCRIPTION
**Pull request by grimpunch**
For https://github.com/OpenRCT2/OpenRCT2/issues/5231,
shows a [HH:MM] timestamp next to chat messages in the clientside.
Defaults to off, and can be toggled with config parameter `timestamp_chat`

### CODE REVIEW

**Comment by grimpunch** [comment]
Changed the config param to be true by default. :+1:

**Comment by marijnvdwerf** [comment]
Can you remove the entire config option?

**Comment by grimpunch** [comment]
Config option has been removed.

File: src/openrct2/network/network.cpp [hunk 1]
```diff
@@ -534,17 +534,25 @@ const char* Network::FormatChat(NetworkPlayer* fromplayer, const char* text)
...
+	struct tm * tmInfo;
+	time(&timer);
+	tmInfo = localtime(&timer);
+	strftime(timeStamp, sizeof(timeStamp), "[%H:%M] ", tmInfo);
 	if (fromplayer) {
 		lineCh = utf8_write_codepoint(lineCh, FORMAT_OUTLINE);
 		lineCh = utf8_write_codepoint(lineCh, FORMAT_BABYBLUE);
-		safe_strcpy(lineCh, (const char *) fromplayer->Name.c_str(), sizeof(formatted) - (lineCh - formatted));
+		safe_strcpy(lineCh, timeStamp, sizeof(formatted) - (lineCh-formatted));
+		safe_strcat(lineCh, (const char *) fromplayer->Name.c_str(), sizeof(formatted) - (lineCh - formatted));
```
Comment (janisozaur) [inline src/openrct2/network/network.cpp]: Can you update the buffer length calculations with the value returned from `strftime` (check for 0 too)?

File: src/openrct2/network/network.cpp [hunk 2]
```diff
@@ -544,8 +544,8 @@ const char* Network::FormatChat(NetworkPlayer* fromplayer, const char* text)
 		lineCh = utf8_write_codepoint(lineCh, FORMAT_OUTLINE);
 		lineCh = utf8_write_codepoint(lineCh, FORMAT_BABYBLUE);
 		safe_strcpy(lineCh, timeStamp, sizeof(formatted) - (lineCh-formatted));
-		safe_strcat(lineCh, (const char *) fromplayer->Name.c_str(), sizeof(formatted) - (lineCh - formatted));
-		safe_strcat(lineCh, ": ", sizeof(formatted) - (lineCh - formatted));
+		safe_strcat(lineCh, (const char *) fromplayer->Name.c_str(), sizeof(formatted) - (lineCh - formatted - timestamp));
```
Comment (janisozaur) [inline src/openrct2/network/network.cpp]: that's no good. You should be using return value from `strftime`

**Comment by grimpunch** [comment]
Provided an alternate solution

**Comment by Gymnasiast** [comment]
Looks good to me 👍 
@janisozaur You can merge this as far as I'm concerned.

File: src/openrct2/network/network.cpp [hunk 3]
```diff
@@ -534,17 +534,23 @@ const char* Network::FormatChat(NetworkPlayer* fromplayer, const char* text)
 	static char formatted[1024];
 	char* lineCh = formatted;
 	formatted[0] = 0;
+	time_t timer;
```
Comment (janisozaur) [inline src/openrct2/network/network.cpp]: This is wasted space & time in case `!fromplayer == true`. Can you move it all down to the conditional block?

Comment (grimpunch) [inline src/openrct2/network/network.cpp]: Yep

**Comment by janisozaur** [comment]
zaznaczenie_213

It doesn't quite work it meant to.

**Comment by grimpunch** [comment]
image

**Comment by grimpunch** [comment]
Aha, it seems the server formats the chat message before sending it to the clients, so the server is responsible for what gets printed...

hmm, guess that means more changes are needed to make this work client side only.

File: src/openrct2/util/util.c [hunk 4]
```diff
@@ -531,3 +532,17 @@ money32 add_clamp_money32(money32 value, money32 value_to_add)
 }
 
 #undef add_clamp_body
+
+/**
+ * strftime wrapper which appends to an existing string.
+ */
+size_t strcatftime(char * buffer, size_t bufferSize, const char * format, const struct tm * tp)
+{
+    size_t stringLen = strlen(buffer);
```
Comment (janisozaur) [inline src/openrct2/util/util.c]: `strnlen`?

Comment (IntelOrca) [inline src/openrct2/util/util.c]: fixed

**Comment by janisozaur** [comment]
One note: the timestamp does not have the outline around it

**Pull request merged**